mrjob/spark/mrjob_spark_harness.py

```diff
@@ -72,17 +72,43 @@ def _run_step(step, step_num, rdd, make_job):
             lambda pairs: mapper_job.map_pairs(pairs, step_num))
         rdd = rdd.map(lambda k_v: m_write(*k_v))
 
+    if step_desc.get('combiner'):
+        combiner_job = make_job('--combiner', '--step-num=%d' % step_num)
+        c_read, c_write = combiner_job.pick_protocols(step_num, 'combiner')
+
+        # run the combiner
+        rdd = rdd.map(c_read)
+
+        def combiner_helper(pairs1, pairs2):
+            if len(pairs1) == len(pairs2) == 1:
+                return list(
+                    combiner_job.combine_pairs(pairs1 + pairs2, step_num),
+                )
+            else:
+                pairs1.extend(pairs2)
+                return pairs1
+
+        rdd = rdd.map(lambda k_v: (k_v[0], k_v))
+        rdd = rdd.combineByKey(
+            createCombiner=lambda k_v: [k_v],
+            mergeValue=lambda k_v_list, k_v: combiner_helper(k_v_list, [k_v]),
+            mergeCombiners=combiner_helper,
+        )
+        rdd = rdd.flatMap(lambda k_to_k_v: k_to_k_v[1])
+
+        rdd = rdd.map(lambda k_v: c_write(*k_v))
```

Comment:
If you did this write step before the `flatMap()`, you could produce the same `key_and_lines` format that the reducer expects. And then you could implement `SORT_VALUES` in the `flatMap()`, like the previous code did.